### DISCUSSION TITLE
Lesson 12: understanding math calculations

### DISCUSSION THREAD

**Discussion by NotNotRama** [Cyfrin/foundry-full-course-cu#469] [Q&A] [answered]
I'm having trouble understanding this function, specifically the `PRECISION` part:

```js
    function getTokenAmountFromUsd(address token, uint256 usdAmountInWei) public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFeed.latestRoundData();
        // 1 ETH = $1000
        // The returned value from CL will be 1000 * 1e8

        return (usdAmountInWei * PRECISION) / (uint256(price) * ADDITIONAL_FEED_PRECISION);
    }

```
If I get this right, the price is in 1e8 units, so for example if the used token was eth, then the price would be 1000e8. But since we need wei, we multiply this by 1e10.

Now, the usd amount is already in wei units. Let's say I passed 10 DSC, `usdAmountInWei` would be 10e18. But why do I need to multiply that amount by 1e18? Isn't that amount already in wei units?

Also, is there any resource to understand the math used in DeFi? So far chatgpt has been great, but not so much for math!

**Comment by bigBagBoogy** [comment]
The math is correct and the function is correct. I struggled a lot too with this. When you continue in the coarse, There will be a HelperConfig where we set the Weth amount to $2000. That already makes the calculation more readable than the $1000/ETH you mention in your remark.
Then we write a test for this: ```testGetTokenAmountFromUsd``` This should hopefully clarify the calculation further I hope.

If you write the function and the test with ```console(log)``` in them like I did in the examples below, then you will get even more insight in the various components.  Just remember to import the console functionality on top with: ```import {console} from "forge-std/Test.sol";```

So If you don't fully get it now, I suggest you just continue the coarse for now until you get to the tests.

Happy coding!


```solidity
function getTokenAmountFromUsd(address token, uint256 usdAmountInWei) public view returns (uint256) {
        console.log("token", token, "usdAmountInWei", usdAmountInWei);
        AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFeed.latestRoundData(); 
        console.log("price", uint256(price));
        console.log("PRECISION", PRECISION);
        console.log("ADDITIONAL_FEED_PRECISION", ADDITIONAL_FEED_PRECISION);
        // $100e18 USD Debt
        // 1 ETH = 2000 USD
        // The returned value from Chainlink will be 2000 * 1e8 (or 2000e18 DSC-WEI)
        // Most USD pairs have 8 decimals, so we will just pretend they all do
        return ((usdAmountInWei * PRECISION) / (uint256(price) * ADDITIONAL_FEED_PRECISION));
    }


////////////// this test belongs in DSCEngineTest.t.sol  ///////

function testGetTokenAmountFromUsd() public {
        // If we want $100 of WETH @ $2000/WETH, that would be 0.05 WETH
        uint256 expectedWeth = 0.05 ether; // this ether here is Ethereum ether
        uint256 amountWeth = dsce.getTokenAmountFromUsd(weth, 100 ether); // 100 ether here is $100 worth of the 1 - 1 stablecoins (called "ether" here) and this is not 100 ETH!!!!
        assertEq(amountWeth, expectedWeth);
        console.log("amountWeth:", amountWeth, "expectedWeth:", expectedWeth);
    }```

**Reply by NotNotRama** [reply]
I already have the `HelperConfig` file and understand most of reasoning behind most of the variables with the exception of this particular part `usdAmountInWei * PRECISION`. The first variable contains the usd value in wei (for example, 10e18), but what I don't get is why I have to multiply it yet again by 1e18 if it's already in wei units.

It's not a deal breaker in understanding solidity I guess, but I'm trying to understand the math behind it.

**Reply by bigBagBoogy** [reply]
Just don't try to use chatGPT for these calculations, because it does not do a good job at it. I thought it would at the very least be a good calculator, but it's not.

If you want you can use ```chisel``` in the command line to do proper calculations.   Just type ```chisel```

```$ chisel
Welcome to Chisel! Type `!help` to show available commands.
➜ 1e18*1000e18
Type: uint
├ Hex: 0x2f050fe938943acc45f65568000000000
└ Decimal: 1000000000000000000000000000000000000000
➜ ```

**Comment by theirrationalone** [comment, marked as answer]
Hope It Helps 👇 (rEaD Comments carefully).

```solidity
    function getTokenAmountFromUsd(address token, uint256 usdAmountInWei) public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFeed.latestRoundData();
        // 1 ETH = $1000
        // The returned value from CL will be 1000 * 1e8

        // I'm Taking usdAmountInWei = 1000000000000000000000 -> 1000_00000000_0000000000
        // It means that we have some tokens worth of 1000000000000000000000 -> 1000_00000000_0000000000 (100018)
        
        // ...

        // usdAmountInWei / uint256(price) * ADDITIONAL_FEED_PRECISION
        // usdAmountInWei: 1000000000000000000000 -> 1000_00000000_00000000 (1000e18) ✅
        // uint256(price): 100000000000 -> 1000_00000000 (1000e8) ✅
        // ADDITIONAL_FEED_PRECISION: 10000000000 -> 1_0000000000(1e10) ✅
        // ...
        // usdAmountInWei / uint256(price) * ADDITIONAL_FEED_PRECISION
        // 1000_00000000_0000000000 / 1000_00000000 * 1_0000000000(1e10)
        // 1000_00000000_0000000000 / 1000_00000000_0000000000
        // 1 <- Tokens, Now we have 1 token that's true though, Remember who called getTokenAmountFromUsd? It was liquidate function.

        // Follow liquidate function from the line where this code below sits on there...
        uint256 collateralTokenAmountToCoverFromUsd = getTokenAmountFromUsd(_collateralTokenAddress, _debtToCover);


        // collateralTokenAmountToCoverFromUsd: 1
        // liquidationBonusAmount: (1 * 50) / 100 -> 0.5 (Means we have 0.5 DSC Token and if we make an exchange of it then we will get 500 USD) <- but floating point values are illegal in solidity. So what we'ld have to do? Let's multiply 0.5 with a common value that we can understand easily and can be understandable(interpretable) to solidity and that's (1000000000000000000 -> 1_00000000_0000000000 ->1e18) <- which is easy to understand, easy to calculate, easy to PRECISIONize.

        // what we have to multiply it with collateralTokenAmountToCoverFromUsd which is 1 why? Because if 0.5 happens then solidity will throw an error : illegal value 0.5 not allowed <- like this maybe?

        // liquidationBonusAmount: (1 * 1e18) * 50 / 100 -> 50_00000000_0000000000 / 100 -> 50_00000000_00000000 -> 0050_00000000_00000000 -> 00500000000000000000
        // 500000000000000000

        // But It's such a headache why? We will have to do this calculation everywhere where we need to get token amount(Number of DSC Tokens) so it would be betttttttttter why not to keep it into getTokenAmountFromUsd in advance.

        // That's why Patrick(SIR) did this, (usdAmountInWei * PRECISION) / (uint256(price) * ADDITIONAL_FEED_PRECISION);
        // because He is The Creator, BTW You can implement a different strategy that makes your life more easier than this and Don't forget to share that with us to make our life easier too.

        // Now let's do this calculation...
        // (1000e18 * 1e18) / (1000e8 * 1e10)
        // 1000e36 / 1000e18
        // 1000e18 Answer. <- PRECISIONED Token Amount (Not any currency's value, It's the bare value of 1 DSC Token in Terms of USD) -> 1 DSC = 1000 USD with PRECISION 1e18.

        return (usdAmountInWei * PRECISION) / (uint256(price) * ADDITIONAL_FEED_PRECISION);
    }

    function liquidate(address _collateralTokenAddress, address _user, uint256 _debtToCover)
        external
        moreThanZero(_debtToCover)
        isValidToken(_collateralTokenAddress)
        nonReentrant
        {
            uint256 userStartingHealthFactor = _healthFactor(_user);

            if (userStartingHealthFactor >= MINIMUM_HEALTH_FACTOR) {
                revert DSCEngine__HealthFactorIsOk();
            }

            uint256 collateralTokenAmountToCoverFromUsd = getTokenAmountFromUsd(_collateralTokenAddress, _debtToCover);

            uint256 liquidationBonusAmount =
                (collateralTokenAmountToCoverFromUsd * LIQUIDATION_BONUS) / LIQUIDATION_PRECISION;

            uint256 totalIncentiveToLiquidateUser = collateralTokenAmountToCoverFromUsd + liquidationBonusAmount;

            _burnDSC(_user, msg.sender, _debtToCover);
            _redeemCollateral(_user, msg.sender, _collateralTokenAddress, totalIncentiveToLiquidateUser);

            uint256 userEndingHealthFactor = _healthFactor(_user);

            if (userEndingHealthFactor <= userStartingHealthFactor) {
                revert DSCEngine__HealthFactorNotImproved();
            }

            _revertIfHealthFactorIsBroken(msg.sender);
        }
```

Sorry for any typo & lang(natural) mispellings....!

**Comment by pierrealex3** [comment]
To add to this, another hint to help understand the involvement of the concept of precision OVERALL - and as a reference, in the context of the referred function of this discussion.

# Things to remember: 

**GOAL**
What we really want out of this function is the USD value of the amount of the ERC20 token specified.  Amount * Price.

PARAMETER -> `uint256 amount`
The amount is in Wei, therefore the precision will always be 18 for this variable. 
For example: 55.55 ETH is represented as uint256 -> 55 55000000 0000000000
This could be viewed as: 55.(55000000 0000000000) // 18 digits

`(,int256 price,,,) = priceFeed.latestRoundData();`

For example, for a price of $1777.56, the variable price would be: 1777 56**000000**
The price is in units of ERC20 for the feeds in scope (ETH or BTC) and its precision is 8.
 If we bring this number back to precision 18 (as is the amount), then the variable price becomes: 1777 56000000 **0000000000**

If we don't want to lose precision on the end result of a multiplication, we must divide with the precision multiplied by itself times (nb of operands -1)
For example, if we have:

# scenario 1
Precision: 3 decimals i.e. 1e3
O1 = operand 1
O2 = operand 2

(O1 * O2) / 1e3


# scenario 2
Precision: 3 decimals i.e. 1e3
O1 = operand 1
O2 = operand 2
O3 = operand 3

(O1 * O2 * O3) / 1e3*1e3 // that is 1e6


In summary, let's say the price in USD is worth $5555,656945.
The returned value (remember, precision 18 is kept in the calculation!) would be represented as uint256: 5555 656945**00 0000000000**

**Reply by EngrPips** [reply]
Interesting, Thanks for this heads-up buddy.
